Support optional float parameters (float?, optional<double>).

gchanan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #31535 Remove temporary test function.
* **#31517 Support optional float parameters (float?, optional<double>).**

This is going to be used by upsample (which currently uses magic values to represent optionals).

For now, we just introduce a fake function for testing (torch._test_optional_float(x)).

Differential Revision: [D19198721](https://our.internmc.facebook.com/intern/diff/D19198721)

gchanan committed:
Support optional float parameters (float?, optional<double>).

This is going to be used by upsample (which currently uses magic values to represent optionals).

For now, we just introduce a fake function for testing (torch._test_optional_float(x)).

[ghstack-poisoned]

jamesr66a commented on torch/csrc/jit/tracer.cpp:
```diff
@@ -444,6 +444,15 @@ void addInputs(Node* n, const char* name /* unused */, const c10::optional<bool>
 void addInputs(Node* n, const char* name, double value) {
   detail::genericAddInput(n, value);
 }
+void addInputs(Node* n, const char* name /* unused */, const c10::optional<double>& value) {
```
Shouldn't `name` be in the comment so that the linter isn't triggered?

gchanan replied:
I just copied `void addInputs(Node* n, const char* name /* unused */, const c10::optional<bool>& value) {` above.

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 872b743:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=872b743dca572ae4cc6d759f435d7dd018bc24a7).

### :detective: 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4035101) pytorch_windows_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=78384939) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=78384939))
<details>
<summary>
<code>CMake Error at torch_cuda_generated_BCECriterion.cu.obj.Release.cmake:281 (message):</code>
</summary>

```
        ]
C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include\thrust/detail/type_traits/pointer_traits.h(257): note: see reference to class template instantiation 'thrust::detail::has_raw_pointer<void *>' being compiled
C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include\thrust/detail/type_traits/pointer_traits.h(114): error C2993: 'T': illegal type for non-type template parameter '__formal'
        with
        [
            T=const void *
        ]
C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include\thrust/detail/type_traits/pointer_traits.h(287): note: see reference to class template instantiation 'thrust::detail::has_raw_pointer<const void *>' being compiled
-- Removing C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THCUNN/./torch_cuda_generated_BCECriterion.cu.obj
"C:/Program Files/CMake/bin/cmake.exe" -E remove C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THCUNN/./torch_cuda_generated_BCECriterion.cu.obj
CMake Error at torch_cuda_generated_BCECriterion.cu.obj.Release.cmake:281 (message):
  Error generating file
  C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THCUNN/./torch_cuda_generated_BCECriterion.cu.obj


s/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/./torch_cuda_generated_THCTensorMathPointwiseDouble.cu.obj -D generated_cubin_file:STRING=C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/./torch_cuda_generated_THCTensorMathPointwiseDouble.cu.obj.cubin.txt -P C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/torch_cuda_generated_THCTensorMathPointwiseDouble.cu.obj.Release.cmake" 
-- Removing C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/./torch_cuda_generated_THCTensorMathPointwiseDouble.cu.obj
"C:/Program Files/CMake/bin/cmake.exe" -E remove C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/./torch_cuda_generated_THCTensorMathPointwiseDouble.cu.obj
-- Generating dependency file: C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/torch_cuda_generated_THCTensorMathPointwiseDouble.cu.obj.NVCC-depend
C:/Users/circleci/project/aten/src/THCUNN -IC:/Users/circleci/project/aten/src/ATen/cuda -IC:/Users/circleci/project/build/caffe2/aten/src -IC:/Users/circleci/project/aten/../third_party/catch/single_include -IC:/Users/circleci/project/aten/src/ATen/.. -IC:/Users/circleci/project/build/caffe2/aten/src/ATen -IC:/Users/circleci/project/c10/cuda/../.. -IC:/Users/circleci/project/c10/../ -IC:/Users/circleci/project/caffe2/../torch/csrc/api -IC:/Users/circleci/project/caffe2/../torch/csrc/api/include
THCTensorMathPointwiseDouble.cu 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

gchanan committed:
Update on "Support optional float parameters (float?, optional<double>)."


This is going to be used by upsample (which currently uses magic values to represent optionals).

For now, we just introduce a fake function for testing (torch._test_optional_float(x)).

Differential Revision: [D19198721](https://our.internmc.facebook.com/intern/diff/D19198721)

[ghstack-poisoned]

gchanan committed:
Update on "Support optional float parameters (float?, optional<double>)."


This is going to be used by upsample (which currently uses magic values to represent optionals).

For now, we just introduce a fake function for testing (torch._test_optional_float(x)).

Differential Revision: [D19198721](https://our.internmc.facebook.com/intern/diff/D19198721)

[ghstack-poisoned]

gchanan committed:
Update on "Support optional float parameters (float?, optional<double>)."


This is going to be used by upsample (which currently uses magic values to represent optionals).

For now, we just introduce a fake function for testing (torch._test_optional_float(x)).

Differential Revision: [D19198721](https://our.internmc.facebook.com/intern/diff/D19198721)

[ghstack-poisoned]

jamesr66a reviewed (approved):
ok boss
